keps/sig-node/20191014-configurable-system-reserved.md

```diff
@@ -0,0 +1,183 @@
+  - sig-node
+reviewers:
+  - "@ConnorDoyle"
+  - "@derekwaynecarr"
+  - "@jianzzha"
+approvers:
+  - TBD
+editor: TBD
+creation-date: 2019-10-14
+last-updated: 2019-10-17
+status: provisional
+see-also:
+  - https://github.com/kubernetes/community/pull/2435
+  - https://github.com/kubernetes/kubernetes/pull/83592
+replaces:
+  - N/A
+superseded-by:
+  - N/A
+---
+
+# Support configuring an exact cpuset as "system-reserved" in Kubelet
+
+## Table of Contents
+
+<!-- toc -->
+- [Release Signoff Checklist](#release-signoff-checklist)
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [User Stories](#user-stories)
+    - [User Story 1](#user-story-1)
+    - [User Story 2](#user-story-2)
+    - [User Story 3](#user-story-3)
+  - [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+    - [Configuration](#configuration)
+    - [Inter-working with other features](#inter-working-with-other-features)
+    - [Proper cpuset configuration for non-static policy use-cases](#proper-cpuset-configuration-for-non-static-policy-use-cases)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+  - [Graduation Criteria](#graduation-criteria)
+  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+- [ ] kubernetes/enhancements issue in release milestone, which links to KEP (this should be a link to the KEP location in kubernetes/enhancements, not the initial KEP PR)
+- [ ] KEP approvers have set the KEP status to `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+## Summary
+Kubelet supports configuring a subset of the Node resources it shall not actively manage via the "system-reserved" flag.
+Currently only CPU, and memory are supported. 
+The CPU Manager -more precisely its static policy- implemented within Kubelet interprets the absolute amount of "CPU shares" coming from this configuration as the "first X number of core IDs I shall not touch".
+X denotes the smallest amount of vCPUs (threads, or physical cores) which are enough to satisfy the request.
+I.e. for system-reserved=1200m CPU Manager excludes CPU cores 0, and 1 form the list of CPU cores usable by K8s workloads.
+This KEP proposes to enhance the format, and the meaning of this flag so that it becomes possible to configure an exact list of CPU cores Kubelet shall entirely exclude from its supervision, under all circumstances.
```

Comment:
Should there be a connection to `--system-reserved-cgroup` option? At the moment, if CPU resources are requested using either `--system-reserved=cpu` or `--reserved-cpus` option, the cpuset belonging to the system reserved cgroup is not set. Only the cpu shares are enforced.

Reply:
correct me if I'm wrong but isn't it the case that if --reserved-cpus is set, **and** the CPU manager policy is set to static then the reserved CPU core IDs are actually removed from the K8s managed pool?

what this KEP is propsing is to take this one step forward, and remove these core IDs from the K8s managed cpuset regardless what is the CPU manager policy (e.g. remove these cores from the "shared pool" even if CPU manager policy is set to "none")

I'm not sure how this would interwork with --system-reserved-cgroup TBH

Reply:
This options works well with kernel parameter --isolcpus and we get pretty good perf number on  containerized dpdk app test. I consider it as a interim solution; in the long term cgroup control 
 (without using isolcpus) might be the right way to go so we need to continue on the cgroup control discussion.

Reply:
I made a PR about how `--reserved-cpus` might work with `--system-reserved-cgroup`, please take a look: https://github.com/kubernetes/kubernetes/pull/87452